### ISSUE TITLE
Cop idea: detect non-linear regexps

### ISSUE THREAD

**Issue by makenowjust** [rubocop/rubocop#11952]
## Is your feature request related to a problem? Please describe.

Since Ruby 3.2, the Regexp engine (Onigmo) has been improved to prevent catastrophic backtracking on matching. Catastrophic backtracking may cause ReDoS (Regular Expression Denial of Service), and this improvement can prevent ReDoS. However, this improvement does not catch all regexps. For example, back-references and subexpression-calls are not supported. Therefore, we want to detect usages of such non-linear features in regexp literals statically.

## Describe the solution you'd like

I suggest adding a new cop `Security/NonLinearRegexp`. This cop can detect usages of non-linear features in a regexp literal statically using `Regexp.linear_time?` also added since Ruby 3.2.

Example:

```ruby
# bad
/("|').+\1/

# good
/foo|bar/
```

## Describe alternatives you've considered

There are some discussion points:

- To analyze a regexp literal having interpolations are hard or impossible. What should we do? (Ignored simply, reporting for safety, or switching its behavior by config?)
- `Regexp.linear_time?` result depends on the Ruby version (its implementation). Therefore, the safety of results can only be guaranteed by the version of Ruby on which Rubocop was run. Is this problematic?

## Additional context

I'm the author of this Regexp engine improvement.

Also, I tried to implement the cop `Security/NonLinearRegexp` in https://github.com/makenowjust/rubocop/tree/non-linear-regexp. When the specification of this cop is fixed, I will open a new pull request.

Thank you.

**Comment by koic** [comment]
I have also implemented a cop for the same purpose locally and ran it as a PoC (This is the so-called "KaigiEffect" of RubyKaigi 😃). But, the reason I ultimately did not open a PR is because I was unsure whether it would be possible to improve the regular expressions while maintaining compatibility. For example, in RuboCop, the following warning is displayed, and it seems some things may not be easy to fix:

```console
$ bundle exec rubocop --only Security/NonLinearRegexp
(snip)

Offenses:

lib/rubocop/config_regeneration.rb:8:21: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
    COMMAND_REGEX = /(?<=`rubocop )(.*?)(?=`)/.freeze
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/correctors/parentheses_corrector.rb:10:24: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        COMMA_REGEXP = /(?<=\))\s*,/.freeze
                       ^^^^^^^^^^^^^
lib/rubocop/cop/exclude_limit.rb:23:27: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
      parameter_name.gsub(/(?<!\A)(?=[A-Z])/, '_').downcase
                          ^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/internal_affairs/node_matcher_directive.rb:112:42: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
          max_pattern_var = pattern.scan(/(?<=%)\d+/).map(&:to_i).max
                                         ^^^^^^^^^^^
lib/rubocop/cop/layout/line_continuation_spacing.rb:69:18: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
            line[/((?<!\s)|\s{2,})\\$/, 0]
                 ^^^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/layout/redundant_line_break.rb:127:19: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
            .gsub(/(["']) *\\\n\s*\1/, '')  # Double or single quote, backslash, then same quote
                  ^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/layout/redundant_line_break.rb:128:19: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
            .gsub(/\n\s*(?=\.\w)/, '')      # Extra space within method chaining
                  ^^^^^^^^^^^^^^^
lib/rubocop/cop/layout/space_inside_array_percent_literal.rb:24:47: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        MULTIPLE_SPACES_BETWEEN_ITEMS_REGEX = /(?:[\S&&[^\\]](?:\\ )*)( {2,})(?=\S)/.freeze
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/layout/space_inside_percent_literal_delimiters.rb:43:21: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        END_REGEX = /(?<!\\)( +)\z/.freeze
                    ^^^^^^^^^^^^^^^
lib/rubocop/cop/lint/duplicate_methods.rb:141:13: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
            /(?:(?<name>.*)::)#<Class:\k<name>>|#<Class:(?<name>.*)>(?:::)?/,
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/lint/interpolation_check.rb:33:25: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
          return unless /(?<!\\)#\{.*\}/.match?(node.source)
                        ^^^^^^^^^^^^^^^^
lib/rubocop/cop/lint/triple_quotes.rb:50:45: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
          opening_quotes = node.source.scan(/(?<=\A)['"]*/)[0]
                                            ^^^^^^^^^^^^^^
lib/rubocop/cop/mixin/line_length_help.rb:82:66: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        if (match = line[end_position..line_length(line)]&.match(/^\S+(?=\s|$)/))
                                                                 ^^^^^^^^^^^^^^
lib/rubocop/cop/style/block_comments.rb:38:79: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
                  contents.source.gsub(/\A/, '# ').gsub(/\n\n/, "\n#\n").gsub(/\n(?=[^#])/, "\n# ")
                                                                              ^^^^^^^^^^^^
lib/rubocop/cop/style/document_dynamic_eval_definition.rb:78:32: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        BLOCK_COMMENT_REGEXP = /^\s*#(?!{)/.freeze
                               ^^^^^^^^^^^^
lib/rubocop/cop/style/document_dynamic_eval_definition.rb:79:26: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        COMMENT_REGEXP = /\s*#(?!{).*/.freeze
                         ^^^^^^^^^^^^^
lib/rubocop/cop/style/numeric_literals.rb:109:58: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
          formatted_int = int_part.abs.to_s.reverse.gsub(/...(?=.)/, '\&_').reverse
                                                         ^^^^^^^^^^
lib/rubocop/cop/style/quoted_symbols.rb:68:12: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
          !/" | (?<!\\)\\[aAbcdefkMnprsStuUxzZ0-7] | \#[@{$]/x.match?(source)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/style/quoted_symbols.rb:113:34: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
          sym_node.source.match?(/\A:?(['"]).*?\1\z/m)
                                 ^^^^^^^^^^^^^^^^^^^^
lib/rubocop/cop/util.rb:137:9: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        /'|(?<! \\) \\{2}* \\ (?![\\"])/x.match?(string)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
spec/project_spec.rb:212:25: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
            entry.match(%r{ ...
                        ^^^
spec/rubocop/cli/suggest_extensions_spec.rb:42:65: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        self.install_suggested = suggestions ? suggestions.scan(/(?<=\* )[a-z0-9_-]+\b/.freeze) : []
                                                                ^^^^^^^^^^^^^^^^^^^^^^^
spec/rubocop/cli/suggest_extensions_spec.rb:45:62: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
        self.load_suggested = suggestions ? suggestions.scan(/(?<=\* )[a-z0-9_-]+\b/.freeze) : []
                                                             ^^^^^^^^^^^^^^^^^^^^^^^
spec/rubocop/cop/range_help_spec.rb:211:19: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
      string.gsub(/^(?!$)/, ' ' * amount)
                  ^^^^^^^^
spec/support/alignment_examples.rb:14:24: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
    name = source.gsub(/\n(?=[a-z ])/, ' <newline> ').gsub(/\s+/, ' ')
                       ^^^^^^^^^^^^^^
tasks/changelog.rb:81:17: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
      str.gsub!(/([A-Z]+)(?=[A-Z][a-z])|([a-z\d])(?=[A-Z])/) do
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tasks/changelog.rb:158:47: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
    @unreleased = parse_release(ss.scan_until(/\n(?=## )/m))
                                              ^^^^^^^^^^^^
tasks/spec_runner.rake:70:24: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
      SUMMARY_REGEXP = /(?<=# SUMMARY BEGIN\n).*(?=\n# SUMMARY END)/m.freeze
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tasks/spec_runner.rake:71:31: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
      FAILURE_OUTPUT_REGEXP = /(?<=# FAILURES BEGIN\n\n).*(?=# FAILURES END)/m.freeze
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
tasks/spec_runner.rake:72:22: C: Security/NonLinearRegexp: Do not use non-linear features in regexp due to the risk of ReDoS.
      RERUN_REGEXP = /(?<=# RERUN BEGIN\n).+(?=\n# RERUN END)/m.freeze
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1502 files inspected, 30 offenses detected
```

So, I'm not confident about how much users would be able to improve the regular expressions that were warned against.

Also, here are my answers:

> - To analyze a regexp literal having interpolations are hard or impossible. What should we do? (Ignored simply, reporting for safety, or switching its behavior by config?)

Yes. Ignoring string interpolation seems reasonable.

> - `Regexp.linear_time?` result depends on the Ruby version (its implementation). Therefore, the safety of results can only be guaranteed by the version of Ruby on which Rubocop was run. Is this problematic?

RuboCop separates static analysis and runtime versions. For instance, I believe there could be inconveniences when statically analyzing `TargetRubyVersion: 3.2` while running on Ruby 3.3. Maybe it can be introduced into core with default disabled, but I'm not sure if this inconsistency is acceptable in RuboCop as a static analyzer.

Therefore, I'm considering that it might be better to make it a separate 3rd party gem rather than being part of the RuboCop core.

@rubocop/rubocop-core I'm not sure, so I'd like to hear other opinions as well.
